go/vt/topo/routing_rules_lock.go

```diff
@@ -18,20 +18,30 @@ package topo
 
 import (
 	"context"
-	"fmt"
 )
 
-// RoutingRulesLock is a wrapper over TopoLock, to serialize updates to routing rules.
-type RoutingRulesLock struct {
-	*TopoLock
+type routingRules struct{}
+
+var _ iTopoLock = (*routingRules)(nil)
+
+func (s *routingRules) Type() string {
+	return "routing_rules"
+}
+
+func (s *routingRules) ResourceName() string {
+	return RoutingRulesPath
+}
+
+func (s *routingRules) Path() string {
+	return RoutingRulesPath
+}
```

Comment:
I don't think this is quite right/ideal. For shards e.g. we have this:
```
type shardLock struct {
	keyspace, shard string
}

func (s *shardLock) ResourceName() string {
	return s.keyspace + "/" + s.shard
}

func (s *shardLock) Path() string {
	return path.Join(KeyspacesPath, s.keyspace, ShardsPath, s.shard)
}
```

So I think the rough equivalent here would be something like this:
```
type routingRules struct{
    subType string
}

func (s *routingRules) ResourceName() string {
	return path.Join(RoutingRulesPath, s.subType)
}

func (s *routingRules) Path() string {
	return path.Join(RoutingRulesPath, subType, CommonRoutingRulesFile) 
}
```

The subTypes AFAIK will eventually be table, shard, and keyspace. With keyspace being the only one that's fully implemented today: https://github.com/vitessio/vitess/blob/0c2856e517ab569476618b0297a175d7c8fcc5da/go/vt/topo/server.go#L69-L97

This make sense, @rohit-nayak-ps and @GuptaManan100 ?

Reply:
That would change the path that we are locking, if we make that change, that will be backward incompatible with the changes that Rohit made previously.

Reply:
I was wrong about what we're currently locking on main: https://github.com/vitessio/vitess/blob/8c978574a3e5e00fbfeb30ad86437f72ea7d5276/go/vt/topo/routing_rules_lock.go#L29-L37

So then the discussion would revolve around whether or not we want to change that. I don't feel too strongly on it. Using separate keys/files for the different types with a shared lock — since they are combined in vtgate for the full set of routing rules in the srvvschema — also makes sense I think.